### PR TITLE
Implement CTPUT in AUTO code logic

### PR DESCRIPTION
**Pull request by wgzintel**
### Details:
 - *Implement CTPUT in AUTO code logic*

### Tickets:
 - *ticket-104956*

### CODE REVIEW

File: src/plugins/auto/auto_schedule.cpp [hunk 1]
```diff
@@ -125,105 +124,181 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
...
             _loadContext[ACTUALDEVICE].deviceInfo.config[CONFIG_KEY(PERFORMANCE_HINT)] =
                 IE::PluginConfigParams::THROUGHPUT;
-            isCumulative = false;
-        } else {
-            // When the hint is ctput and there are more than one device, the MULTI logic is used
-            std::string deviceName = "MULTI:";
+        } else if (validDevices.size() > 1) {
+            _loadContext[ACTUALDEVICE].isEnabled = false;
+            _autoSContext->_devicePriorities.clear();
+            for (auto& validDevice : validDevices) {
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:135]:
```suggestion
            std::copy(std::begin(validDevices), std::end(validDevices), std::back_inserter(_autoSContext->_devicePriorities));
```

File: src/plugins/auto/auto_schedule.cpp [hunk 2]
```diff
@@ -125,105 +124,181 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
...
+            }
+            // Total number of devices in CTPUT
+            _nCTPUTDeviceNums = validDevices.size();
+            DeviceInformation cpuDeviceInformation;
+            const auto CPUIter =
+                std::find_if(validDevices.begin(), validDevices.end(), [=](const DeviceInformation& d) -> bool {
+                    return d.deviceName.find("CPU") != std::string::npos;
+                });
+            if (CPUIter != validDevices.end()) {
+                if (CPUIter->config.find(ov::affinity.name()) == CPUIter->config.end()) {
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:146]: insert function will already check the existence of key right? why still need the if here

File: src/plugins/auto/auto_schedule.cpp [hunk 3]
```diff
@@ -125,105 +124,181 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
...
+                _loadContext[CPU].workName = "CPU_HELP";
+                LOG_INFO_TAG("will load CPU for accelerator");
+            } else {
+                _loadContext[CPU].isEnabled = false;
+            }
+        }
+        // initialize the rest members of load context
+        for (int i = 0; i < CONTEXTNUM; i++) {
+            if (_loadContext[i].isEnabled) {
+                _loadContext[i].future = _loadContext[i].promise.get_future();
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:279]: future not needed in ctput mode

File: src/plugins/auto/auto_schedule.cpp [hunk 4]
```diff
@@ -433,6 +526,21 @@ void AutoSchedule::WaitFirstNetworkReady() {
             LOG_ERROR_TAG("load failed, %s", _loadContext[i].errMessage.c_str());
         }
     }
+    // devices loaded successfully in CTPUT
+    if (_pCTPUTLoadContext) {
+        int nLoadSucNums = 0;
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:531]: no need of future check here

File: src/plugins/auto/auto_schedule.cpp [hunk 5]
```diff
@@ -459,15 +567,21 @@ bool AutoSchedule::ScheduleToWorkerInferRequest(IE::Task inferPipelineTask, Devi
         }
         devices.push_back(_loadContext[ACTUALDEVICE].deviceInfo);
     } else {
-        // _acceleratorDevice could be the same as _cpuDevice, such as AUTO:CPU
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:462]: the preferred device code path need to be updated with ctput also

File: src/plugins/auto/auto_schedule.hpp [hunk 6]
```diff
@@ -68,6 +69,7 @@ class AutoSchedule : public MultiSchedule {
     bool                                     _exitFlag = {false};
     size_t                                   _cpuHelpInferCount = 0;
     AutoScheduleContext::Ptr                 _autoSContext;
+    size_t                                   _nCTPUTDeviceNums = 0;
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.hpp:72]:
```suggestion
    size_t                                   _nCTputDeviceNums = 0;
```

File: src/plugins/auto/auto_schedule.cpp [hunk 7]
```diff
@@ -125,105 +124,181 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
...
+            _loadContext[ACTUALDEVICE].isEnabled = false;
+            _autoSContext->_devicePriorities.clear();
+            for (auto& validDevice : validDevices) {
+                _autoSContext->_devicePriorities.push_back(validDevice);
+            }
+            // Total number of devices in CTPUT
+            _nCTPUTDeviceNums = validDevices.size();
+            DeviceInformation cpuDeviceInformation;
+            const auto CPUIter =
+                std::find_if(validDevices.begin(), validDevices.end(), [=](const DeviceInformation& d) -> bool {
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:142]:
```suggestion
                std::find_if(validDevices.begin(), validDevices.end(), [](const DeviceInformation& d) -> bool {
```

File: src/plugins/auto/auto_schedule.cpp [hunk 8]
```diff
@@ -125,105 +124,181 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
...
+                             [&](const DeviceInformation& d) -> bool {
+                                 return d.deviceName.find(failedDeviceName) != std::string::npos;
+                             });
+            // Remove failed device from _devicePriorities
+            if (DeviceIter != _autoSContext->_devicePriorities.end()) {
+                _autoSContext->_devicePriorities.erase(DeviceIter);
+            }
+            // Remove failed device from ov::device::priorities in config
+            auto it_prior = _autoSContext->_config.find(ov::device::priorities.name());
+            if (it_prior != _autoSContext->_config.end()) {
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:233]: why do we have to remove the config here, this is set by user right?

File: src/plugins/auto/auto_schedule.cpp [hunk 9]
```diff
@@ -459,15 +567,21 @@ bool AutoSchedule::ScheduleToWorkerInferRequest(IE::Task inferPipelineTask, Devi
         }
         devices.push_back(_loadContext[ACTUALDEVICE].deviceInfo);
     } else {
-        // _acceleratorDevice could be the same as _cpuDevice, such as AUTO:CPU
-        if (_loadContext[ACTUALDEVICE].isAlready) {
-            devices.push_back(_loadContext[ACTUALDEVICE].deviceInfo);
+        if (_pCTPUTLoadContext) {
+            for (size_t i = 0; i < _autoSContext->_devicePriorities.size(); i++) {
+                devices.push_back(_autoSContext->_devicePriorities[i]);
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:572]: @wangleis do we need to support runtime change of device priority list with auto ctput?

Comment (songbell) [reply src/plugins/auto/auto_schedule.cpp:572]: if yes, we need to handle this line thread-safe.

Comment (wangleis) [reply src/plugins/auto/auto_schedule.cpp:572]: @songbell  Runtime priority change is planned with round robin policy in following PR after 2023.0.

Comment (songbell) [reply src/plugins/auto/auto_schedule.cpp:572]: see, if then, I suppose related docs which involves this functionality needs to be updated?

File: src/tests/unit/auto/auto_ctput_test.cpp [hunk 10]
```diff
@@ -220,14 +220,20 @@ TEST_P(LoadNetworkWithCTPUTMockTest, CTPUTSingleDevLogicTest) {
...
         }
         config.insert({InferenceEngine::MultiDeviceConfigParams::KEY_MULTI_DEVICE_PRIORITIES, targetDev});
-        // Call MULTI logic
+        // no MULTI logic to be called
         EXPECT_CALL(*core,
                     LoadNetwork(::testing::Matcher<const InferenceEngine::CNNNetwork&>(_),
                                 ::testing::Matcher<const std::string&>("MULTI:" + targetDev),
                                 ::testing::Matcher<const std::map<std::string, std::string>&>(_)))
-            .Times(1);
+            .Times(0);
```
Comment (songbell) [inline src/tests/unit/auto/auto_ctput_test.cpp:236]: something minor: we can remove the expect_call of multi here.

File: src/plugins/auto/auto_schedule.cpp [hunk 11]
```diff
@@ -125,105 +124,181 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
...
+        LOG_INFO_TAG("select device:%s", _loadContext[ACTUALDEVICE].deviceInfo.deviceName.c_str());
+        bool isActualDevCPU = _loadContext[ACTUALDEVICE].deviceInfo.deviceName.find("CPU") != std::string::npos;
+        // if Actual device is CPU or perf_hint is cumulative, disabled _loadContext[CPU], only use
+        // _loadContext[ACTUALDEVICE]
+        if (isActualDevCPU || !_autoSContext->_startupfallback) {
             _loadContext[CPU].isEnabled = false;
+        } else {
+            const auto CPUIter = std::find_if(_autoSContext->_devicePriorities.begin(),
+                                              _autoSContext->_devicePriorities.end(),
+                                              [=](const DeviceInformation& d) -> bool {
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:262]:
```suggestion
                                              [](const DeviceInformation& d) -> bool {
```

File: src/plugins/auto/auto_schedule.cpp (hunk 2 above)
Comment (wgzintel) [reply src/plugins/auto/auto_schedule.cpp:146]: Check here to see if the user has set affinity, if not set, it will be set to CORE by default.

File: src/plugins/auto/auto_schedule.cpp (hunk 8 above)
Comment (wgzintel) [reply src/plugins/auto/auto_schedule.cpp:233]: AutoExecutableNetwork::GetMetric(ov::device::priorities.name())  does not return the  fail loaded device, so remove failed  device from ov::device::priorities.

File: src/plugins/auto/auto_schedule.cpp (hunk 11 above)
Comment (wgzintel) [reply src/plugins/auto/auto_schedule.cpp:262]: Updated.

File: src/plugins/auto/auto_schedule.cpp [hunk 12]
```diff
@@ -302,14 +377,32 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
...
-        // only one device need to load network, do not need to load it async
-        _loadContext[ACTUALDEVICE].task();
-        _passthroughExeNet = _loadContext[ACTUALDEVICE].executableNetwork;
+        if (_pCTPUTLoadContext) {
+            _executor = _autoSContext->_plugin->executorManager()->getIdleCPUStreamsExecutor(IStreamsExecutor::Config{
+                "CTPUTDeviceAsyncLoad",
+                static_cast<int>(std::thread::hardware_concurrency()) /* max possible #streams*/,
+                0 /*default threads per stream, workaround for ticket 62376*/,
+                IStreamsExecutor::ThreadBindingType::NONE});
+            // Load other devices fist
```
[the remark this answers is not in the document]
Comment (wgzintel) [reply src/plugins/auto/auto_schedule.cpp:386]: Updated

File: src/plugins/auto/auto_schedule.hpp (hunk 6 above)
Comment (wgzintel) [reply src/plugins/auto/auto_schedule.hpp:72]: Updated

File: src/tests/unit/auto/auto_ctput_test.cpp (hunk 10 above)
Comment (wgzintel) [reply src/tests/unit/auto/auto_ctput_test.cpp:236]: Updated, remove ecpect_call

File: src/plugins/auto/auto_schedule.cpp (hunk 3 above)
Comment (wgzintel) [reply src/plugins/auto/auto_schedule.cpp:279]: Updated.

File: src/plugins/auto/auto_schedule.cpp (hunk 7 above)
Comment (wgzintel) [reply src/plugins/auto/auto_schedule.cpp:142]: Updated

File: src/plugins/auto/auto_schedule.cpp (hunk 1 above)
Comment (wgzintel) [reply src/plugins/auto/auto_schedule.cpp:135]: Updated

File: src/plugins/auto/auto_executable_network.cpp [hunk 13]
```diff
@@ -183,6 +187,9 @@ IE::Parameter AutoExecutableNetwork::GetMetric(const std::string& name) const {
         std::lock_guard<std::mutex> lock(_autoSContext->_confMutex);
         if (_autoSchedule->_loadContext[CPU].isEnabled && _autoSchedule->_loadContext[CPU].isAlready)
             return _autoSchedule->_loadContext[CPU].executableNetwork->GetMetric(name);
+        if (_autoSchedule->_pCTPUTLoadContext) {
```
[the remark this answers is not in the document]
Comment (wgzintel) [reply src/plugins/auto/auto_executable_network.cpp:190]: Updated

File: src/plugins/auto/auto_schedule.cpp (hunk 3 above)
Comment (wgzintel) [reply src/plugins/auto/auto_schedule.cpp:279]: Updated in line 291

File: src/plugins/auto/auto_schedule.cpp (hunk 2 above)
Comment (songbell) [reply src/plugins/auto/auto_schedule.cpp:146]: I understand, but I mean the insert function will not update the affinity key if already existed.

CPUIter->config.insert(
                        {ov::affinity.name(), ov::affinity(ov::Affinity::CORE).second.as<std::string>()});
// insert the affinity if not in config, else(if this key already in config), do nothing.

File: src/plugins/auto/auto_schedule.cpp [hunk 14]
```diff
@@ -125,105 +124,180 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
...
+            const auto CPUIter =
+                std::find_if(validDevices.begin(), validDevices.end(), [](const DeviceInformation& d) -> bool {
+                    return d.deviceName.find("CPU") != std::string::npos;
+                });
+            if (CPUIter != validDevices.end()) {
+                if (CPUIter->config.find(ov::affinity.name()) == CPUIter->config.end()) {
+                    LOG_DEBUG_TAG("set affinity to CORE for CPU");
+                    // no user set affinity then set CORE to CPU
+                    CPUIter->config.insert(
+                        {ov::affinity.name(), ov::affinity(ov::Affinity::CORE).second.as<std::string>()});
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:150]:
```suggestion
                        {ov::affinity.name(), ov::Any(ov::Affinity::CORE).as<std::string>()});
```

File: src/plugins/auto/auto_schedule.cpp [hunk 15]
```diff
@@ -125,105 +124,180 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
...
+            }
+            // Generate contexts for loading each device
+            _pCTPUTLoadContext.reset(new AutoLoadContext[_nCTputDeviceNums]);
+            int idx = 0;
             for (auto& device : validDevices) {
-                deviceName += device.deviceName;
-                deviceName += ((device.deviceName == validDevices.back().deviceName) ? "" : ",");
+                _pCTPUTLoadContext[idx].deviceInfo = device;
+                _pCTPUTLoadContext[idx].deviceInfo.config[CONFIG_KEY(PERFORMANCE_HINT)] = IE::PluginConfigParams::THROUGHPUT;
+                idx++;
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:161]: @wangleis do we need to support latency mode with auto_ctput?

Comment (wangleis) [reply src/plugins/auto/auto_schedule.cpp:161]: @songbell Will not support latency mode in 2023.0 for same behavior between MULTI and AUTO CTPUT. Need customer use case to clarify the requirement first.

Comment (wangleis) [reply src/plugins/auto/auto_schedule.cpp:161]: Need clarification in document for this change in 2023.0

File: src/plugins/auto/auto_schedule.cpp (hunk 5 above)
Comment (wgzintel) [reply src/plugins/auto/auto_schedule.cpp:462]: Updated.

File: src/plugins/auto/auto_schedule.cpp (hunk 2 above)
Comment (wgzintel) [reply src/plugins/auto/auto_schedule.cpp]: Updated

File: src/plugins/auto/auto_schedule.cpp (hunk 14 above)
Comment (wgzintel) [reply src/plugins/auto/auto_schedule.cpp]: Updated

File: src/plugins/auto/auto_schedule.cpp (hunk 15 above)
Comment (peterchen-intel) [reply src/plugins/auto/auto_schedule.cpp:157]: What is the behavior if user -d MULTI:GPU,CPU -hint LATENCY?  Will threw exception?   We can add the expected behavior in CVS-104939, and document in  CVS-105015

File: src/plugins/auto/common.hpp [hunk 16]
```diff
@@ -157,6 +157,7 @@ class AutoScheduleContext : public MultiScheduleContext {
     IE::CNNNetwork              _network;
     std::string                 _strDevices;
     unsigned int                _modelPriority = 0;
+    unsigned int                _ctputOtimalNums = 0;
```
Comment (songbell) [inline src/plugins/auto/common.hpp:160]:
```suggestion
    unsigned int                _ctputOptimalNums = 0;
```

File: src/plugins/auto/auto_executable_network.cpp [hunk 17]
```diff
@@ -162,14 +165,15 @@ IE::Parameter AutoExecutableNetwork::GetMetric(const std::string& name) const {
             if (ExeDevicesString == "CPU_HELP")
                 ExeDevicesString = "(CPU)";
             exeDevices.push_back(ExeDevicesString);
-            execution_devices = decltype(ov::execution_devices)::value_type {exeDevices};
+            execution_devices = decltype(ov::execution_devices)::value_type{exeDevices};
```
Comment (songbell) [inline src/plugins/auto/auto_executable_network.cpp:168]:
```suggestion
            execution_devices = decltype(ov::execution_devices)::value_type {exeDevices};
```

File: src/plugins/auto/auto_executable_network.cpp [hunk 18]
```diff
@@ -162,14 +165,15 @@ IE::Parameter AutoExecutableNetwork::GetMetric(const std::string& name) const {
...
-                execution_devices = _autoSchedule->_loadContext[ACTUALDEVICE].executableNetwork->GetMetric(name);
-            } catch(const IE::Exception&) {
-                GetExecutionDevices(_autoSchedule->_loadContext[ACTUALDEVICE].workName);
+        if (_autoSchedule->_pCTPUTLoadContext) {
+            std::vector<std::string> exeDevices = {};
+            std::lock_guard<std::mutex> lock(_autoSContext->_confMutex);
+            for (auto n : _autoSContext->_devicePriorities) {
+                exeDevices.push_back(n.deviceName);
             }
+            execution_devices = decltype(ov::execution_devices)::value_type{exeDevices};
```
Comment (songbell) [inline src/plugins/auto/auto_executable_network.cpp:176]:
```suggestion
            execution_devices = decltype(ov::execution_devices)::value_type {exeDevices};
```

File: src/plugins/auto/auto_schedule.cpp [hunk 19]
```diff
@@ -217,107 +216,178 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
...
-            // can not change _autoSContext->_performanceHint to THROUGHPUT, because GetMetric needs to return CTPUT
             _loadContext[ACTUALDEVICE].deviceInfo = validDevices.front();
             _loadContext[ACTUALDEVICE].deviceInfo.config[CONFIG_KEY(PERFORMANCE_HINT)] =
                 IE::PluginConfigParams::THROUGHPUT;
-            isCumulative = false;
-        } else {
-            // When the hint is ctput and there are more than one device, the MULTI logic is used
-            std::string deviceName = "MULTI:";
+        } else if (validDevices.size() > 1) {
+            _loadContext[ACTUALDEVICE].isEnabled = false;
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:225]: @WeldonWangwang please review the fallback part

File: src/plugins/auto/auto_schedule.cpp [hunk 20]
```diff
@@ -217,107 +216,178 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
...
-                deviceName += device.deviceName;
-                deviceName += ((device.deviceName == validDevices.back().deviceName) ? "" : ",");
+                _pCTPUTLoadContext[idx].deviceInfo = device;
+                _pCTPUTLoadContext[idx].deviceInfo.config[CONFIG_KEY(PERFORMANCE_HINT)] = IE::PluginConfigParams::THROUGHPUT;
+                idx++;
+            }
+            // If there is a CPU, the CPU is loaded last
+            if (cpuDeviceInformation.deviceName.find("CPU") != std::string::npos) {
+                _pCTPUTLoadContext[idx].deviceInfo = cpuDeviceInformation;
+                _pCTPUTLoadContext[idx].deviceInfo.config[CONFIG_KEY(PERFORMANCE_HINT)] = IE::PluginConfigParams::THROUGHPUT;
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:254]:
```suggestion
            DeviceInformation cpuDeviceInformation;
            for (auto& device : validDevices) {
                if (device.deviceName.find("CPU") == std::string::npos) {
                    _pCTPUTLoadContext[idx].deviceInfo = device;
                    _pCTPUTLoadContext[idx].deviceInfo.config[CONFIG_KEY(PERFORMANCE_HINT)] = IE::PluginConfigParams::THROUGHPUT;
                    idx++;
                } else {                    
                    cpuDeviceInformation = device;
                    cpuDeviceInformation.config.insert({ov::affinity.name(), ov::Any(ov::Affinity::CORE).as<std::string>()});
                }
            }
            if (!cpuDeviceInformation.deviceName.empty()) {
                _pCTPUTLoadContext[idx].deviceInfo = cpuDeviceInformation;
                _pCTPUTLoadContext[idx].deviceInfo.config[CONFIG_KEY(PERFORMANCE_HINT)] = IE::PluginConfigParams::THROUGHPUT;
            }
```

File: src/plugins/auto/auto_executable_network.cpp (hunk 17 above)
Comment (wgzintel) [reply src/plugins/auto/auto_executable_network.cpp]: Updated

File: src/plugins/auto/auto_executable_network.cpp (hunk 18 above)
Comment (wgzintel) [reply src/plugins/auto/auto_executable_network.cpp]: Updated

File: src/tests/unit/auto/auto_runtime_fallback_test.cpp [hunk 21]
```diff
@@ -362,3 +364,93 @@ const std::vector<ConfigParams> testConfigs = {
 INSTANTIATE_TEST_SUITE_P(smoke_AutoRuntimeFallback, AutoRuntimeFallback,
                 ::testing::ValuesIn(testConfigs),
            AutoRuntimeFallback::getTestCaseName);
+
+TEST_P(AutoCTPUTRuntimeFallback, ctputDeviceInferFailTest) {
+    std::string targetDev;
+    std::vector<std::tuple<std::string, bool>> targetDevices; //std::tuple<deviceName, will infer throw exception>
```
Comment (songbell) [inline src/tests/unit/auto/auto_runtime_fallback_test.cpp:370]: can we put the general param initialization to base class SetUp() function?

File: src/plugins/auto/auto_schedule.cpp (hunk 5 above)
Comment (songbell) [reply src/plugins/auto/auto_schedule.cpp:571]: kind of redundant with line#731, probably we just need to avoid waitActual in ctput mode.
btw, with remote blob and runtime fall back enabled, we should break out the runtime fallback loop because the fallback device is never able to deal with other device's remote memory. Is this been implemented?

File: src/plugins/auto/auto_schedule.cpp [hunk 22]
```diff
@@ -217,107 +237,178 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
...
             }
-            _loadContext[ACTUALDEVICE].deviceInfo.deviceName = deviceName;
-            _loadContext[ACTUALDEVICE].deviceInfo.config[CONFIG_KEY(PERFORMANCE_HINT)] =
-                InferenceEngine::PluginConfigParams::CUMULATIVE_THROUGHPUT;
-            _loadContext[ACTUALDEVICE].deviceInfo.config[CONFIG_KEY(PERF_COUNT)] =
-                _autoSContext->_needPerfCounters ? InferenceEngine::PluginConfigParams::YES
-                                                 : InferenceEngine::PluginConfigParams::NO;
-            if (_autoSContext->_bindBuffer)
```
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:241]: can we add bind schedule support for auto ctput? or do you plan in a separate PR?

File: src/plugins/auto/auto_schedule.cpp (hunk 15 above)
Comment (songbell) [reply src/plugins/auto/auto_schedule.cpp:270]: I think this is not a huge change for current code, -d MULTI with latency, we just switch multi->auto ctput in auto plugin(set the flag in autoSContext), and just carry on other configs from set_property or compile_model, and pass to load lamdas in AUTO.
so shall we update the logic to support latency, or update the document (and communicate the changes)? what do you think? @peterchen-intel

File: src/plugins/auto/common.hpp (hunk 16 above)
Comment (wgzintel) [reply src/plugins/auto/common.hpp]: Updated

File: src/plugins/auto/auto_schedule.cpp [hunk 23]
```diff
@@ -411,14 +499,32 @@ void AutoSchedule::init(const ScheduleContext::Ptr& sContext) {
         }
         _loadContext[ACTUALDEVICE].task();
     } else {
-        // only one device need to load network, do not need to load it async
-        _loadContext[ACTUALDEVICE].task();
-        _passthroughExeNet = _loadContext[ACTUALDEVICE].executableNetwork;
+        if (_pCTPUTLoadContext) {
```
Comment (WeldonWangwang) [inline src/plugins/auto/auto_schedule.cpp:502]:
```suggestion
        if (_pCTPUTLoadContext) {
            for (auto&& device : _autoSContext->_devicePriorities) {
                // initialize containers before run async task
                _idleWorkerRequests[device.deviceName];
                _workerRequests[device.deviceName];
                _inferPipelineTasksDeviceSpecific[device.deviceName] = nullptr;
            }
```

Comment (wgzintel) [reply src/plugins/auto/auto_schedule.cpp:502]: Updated

**Review by songbell — APPROVED** [review]

**Review by wangleis — APPROVED** [review]
